File: backend/src/providers/orderProvider.ts
import { prisma } from '../lib/prisma.js';
import { CreateOrderBodyType } from '../schemas/order.js';
import { BaseSearchRequestType } from '../schemas/search.js';
import parseFilterString from '../utils/filterParser.js';

/**
 * Data Access Layer — Order
 */
export const orderProvider = {
  /** Find a single order by id. */
  async findById(id: bigint) {
    return prisma.order.findFirst({
      where: { id, deleted: false },
      include: {
        items: {
          where: { deleted: false },
          include: {
            menu_item: {
              include: {
                menu_item_images: {
                  where: { is_primary: true },
                  orderBy: [{ display_order: 'asc' }],
                  take: 1,
                },
              },
            },
          },
        },
      },
    });
  },

  /** Return all active orders, optionally filtered by status. */
  async findAll(request: BaseSearchRequestType) {
    const where = parseFilterString(request.search, {
      allowedFields: [
        'id',
        'customer_id',
        'ticket_number',
        'table_number',
        'status',
        'wait_level',
        'is_paid',
        'payment_method',
        'paid_at',
      ],
      fieldTypes: {
        id: 'number',
        customer_id: 'number',
        ticket_number: 'string',
        table_number: 'string',
        status: 'string',
        wait_level: 'string',
        is_paid: 'boolean',
        payment_method: 'string',
        paid_at: 'date',
      },
      defaultSearchFields: ['status'],
      defaultDeleted: true,
    });

    return prisma.order.findMany({
      where,
      orderBy: { id: request.order },
      take: request.limit + 1,
      ...(request.cursor ? { cursor: { id: request.cursor }, skip: 1 } : {}),
      include: {
        items: {
          where: { deleted: false },
          include: {
            menu_item: {
              include: {
                menu_item_images: {
                  where: { is_primary: true },
                  orderBy: [{ display_order: 'asc' }],
                  take: 1,
                },
              },
            },
          },
        },
      },
    });
  },

  /** Insert a new order with its line items in a single transaction. */
  async create(data: CreateOrderBodyType, ticketNumber: string) {
    return prisma.order.create({
      data: {
        ticket_number: ticketNumber,
        table_number: data.tableNumber,
        status: 'Received',
        total: data.total,
        customer_id: data.customerId,
        items: {
          create: data.items.map((item, idx) => ({
            menu_item_id: item.menuItemId,
            name_at_order: item.nameAtOrder,
            qty: item.qty,
            price_at_order: item.priceAtOrder,
            modifications: item.modifications,
          })),
        },
      },
      include: { items: true },
    });
  },

  /** Return all orders placed by a specific customer, newest first. */
  async findByCustomerId(request: BaseSearchRequestType) {
    const where = parseFilterString(request.search, {
      allowedFields: [
        'id',
        'customer_id',
        'ticket_number',
        'table_number',
        'status',
        'wait_level',
        'is_paid',
        'payment_method',
        'paid_at',
      ],
      fieldTypes: {
        id: 'number',
        customer_id: 'number',
        ticket_number: 'string',
        table_number: 'string',
        status: 'string',
        wait_level: 'string',
        is_paid: 'boolean',
        payment_method: 'string',
        paid_at: 'date',
      },
      defaultSearchFields: ['status'],
      defaultDeleted: true,
    });

    // If controller/service supplied customer_id directly, apply it as a filter.
    const suppliedCustomerId = (request as any).customer_id;
    if (suppliedCustomerId != null) {
      // Prisma expects BigInt for the customer_id column
      (where as any).customer_id = BigInt(suppliedCustomerId);
    }

    return prisma.order.findMany({
      where,
      orderBy: { id: request.order },
      take: request.limit + 1,
      ...(request.cursor ? { cursor: { id: request.cursor }, skip: 1 } : {}),
      include: {
        items: {
          include: {
            menu_item: {
              include: {
                menu_item_images: {
                  where: { is_primary: true },
                  orderBy: [{ display_order: 'asc' }],
                  take: 1,
                },
              },
            },
          },
        },
      },
    });
  },

  /** Update only the `status` field of an order. */
  async updateStatus(id: bigint, status: string) {
    return prisma.order.update({
      where: { id },
      data: { status },
    });
  },

  /** Mark an order as paid and persist payment metadata. */
  async markAsPaid(id: bigint, paymentMethod: string, paidAt: Date) {
    await prisma.$executeRaw`
      UPDATE orders
      SET is_paid = TRUE,
          payment_method = ${paymentMethod},
          paid_at = ${paidAt}
      WHERE id = ${id}
    `;

    return prisma.order.findFirst({ where: { id, deleted: false } });
  },
};

File: backend/app/providers/order_provider.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, text
from sqlalchemy.orm import selectinload
from typing import Optional, List
from datetime import datetime
from app.models.order import Order
from app.models.order_item import OrderItem
from app.models.menu_item import MenuItem
from app.models.menu_item_image import MenuItemImage

class OrderProvider:
    async def find_by_id(self, db: AsyncSession, order_id: int) -> Optional[Order]:
        result = await db.execute(
            select(Order)
            .options(
                selectinload(Order.items).selectinload(OrderItem.menu_item).selectinload(MenuItem.menu_item_images)
            )
            .where(Order.id == order_id, Order.deleted == False)
        )
        return result.scalar_one_or_none()

    async def find_all(self, db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None):
        from sqlalchemy import asc, desc
        stmt = (
            select(Order)
            .options(selectinload(Order.items).selectinload(OrderItem.menu_item).selectinload(MenuItem.menu_item_images))
            .where(Order.deleted == False)
        )
        if search:
            stmt = stmt.where(Order.status.ilike(f"%{search}%") | Order.ticket_number.ilike(f"%{search}%"))
        if cursor:
            if order == "desc":
                stmt = stmt.where(Order.id < cursor)
            else:
                stmt = stmt.where(Order.id > cursor)
        stmt = stmt.order_by(desc(Order.id) if order == "desc" else asc(Order.id))
        stmt = stmt.limit(limit + 1)
        result = await db.execute(stmt)
        return result.scalars().all()

    async def find_by_customer_id(self, db: AsyncSession, customer_id: int, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None):
        from sqlalchemy import asc, desc
        stmt = (
            select(Order)
            .options(selectinload(Order.items).selectinload(OrderItem.menu_item).selectinload(MenuItem.menu_item_images))
            .where(Order.deleted == False, Order.customer_id == customer_id)
        )
        if search:
            stmt = stmt.where(Order.status.ilike(f"%{search}%"))
        if cursor:
            if order == "desc":
                stmt = stmt.where(Order.id < cursor)
            else:
                stmt = stmt.where(Order.id > cursor)
        stmt = stmt.order_by(desc(Order.id) if order == "desc" else asc(Order.id))
        stmt = stmt.limit(limit + 1)
        result = await db.execute(stmt)
        return result.scalars().all()

    async def create(self, db: AsyncSession, table_number: str, ticket_number: str, customer_id: Optional[int], total: int, items_data: list) -> Order:
        order = Order(
            ticket_number=ticket_number,
            table_number=table_number,
            status="Received",
            total=total,
            customer_id=customer_id,
        )
        db.add(order)
        await db.flush()  # get order.id
        for item in items_data:
            order_item = OrderItem(
                order_id=order.id,
                menu_item_id=item["menuItemId"],
                name_at_order=item["nameAtOrder"],
                qty=item["qty"],
                price_at_order=item["priceAtOrder"],
                modifications=item.get("modifications", []),
            )
            db.add(order_item)
        await db.commit()
        await db.refresh(order)
        return order

    async def update_status(self, db: AsyncSession, order_id: int, status: str) -> Optional[Order]:
        await db.execute(update(Order).where(Order.id == order_id).values(status=status))
        await db.commit()
        return await self.find_by_id(db, order_id)

    async def mark_as_paid(self, db: AsyncSession, order_id: int, payment_method: str, paid_at: datetime) -> Optional[Order]:
        await db.execute(
            update(Order).where(Order.id == order_id).values(
                is_paid=True,
                payment_method=payment_method,
                paid_at=paid_at,
            )
        )
        await db.commit()
        return await self.find_by_id(db, order_id)

order_provider = OrderProvider()
